### PR TITLE
LMBlock

### PR DESCRIPTION
**Pull request by KexinFeng**
The local testing model file can be found in https://github.com/deepjavalibrary/djl/pull/2547. But it is runnable with huggingface model too.

This is a series of PRs. The updates will be done on the latest one: https://github.com/deepjavalibrary/djl-serving/pull/743

### CODE REVIEW

File: engines/python/setup/djl_python/scheduler/lm_block.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+
+class HuggingfaceGTP2Block(LMBlock):
+
+    def __init__(self, model_urls: List[str], config: Dict):
+        super(HuggingfaceGTP2Block, self).__init__()
+        self.config = {
+            'use_cache': config.get('use_cache', True),
+            'token_type_ids': config.get('token_type_ids', None),
+            'return_dict': config.get('return_dict', False),
```
Comment (sindhuvahinis) [inline engines/python/setup/djl_python/scheduler/lm_block.py:73]: why return_dict is set to False by default? With return_dict we can easily look up logits, past_key_values etc right?

File: engines/python/setup/djl_python/scheduler/lm_block.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    def __init__(self, model_urls: List[str], gpt_config: GPTConfig):
+        super(OrtLMBlock, self).__init__()
+        self.blocks = [torch.jit.load(url) for url in model_urls]
+        self.gpt_config = gpt_config
+
+    def forward(self, input: List[torch.tensor], past_key_values):
+        raise ("Not implemented yet")
+
+
+class HuggingfaceGTP2Block(LMBlock):
```
Comment (lanking520) [inline engines/python/setup/djl_python/scheduler/lm_block.py:66]: why we want to rewrite the GPT2 Block?

Comment (sindhuvahinis) [reply engines/python/setup/djl_python/scheduler/lm_block.py:66]: I think Qing has a point. 
Other backengines like DS, model.forward is called this way
```
model = deepspeed.init_inference(GPT2LMHeadModel.pretrained("gpt2"), config) # returns the DS engine wrapped model
model.forward(input_ids, attention_mask, ...) # this one internally calls the HF model.forward anyway
```
https://github.com/microsoft/DeepSpeed/blob/d1c3c0df5318cbcb2b221d38d4127dd566f4984d/deepspeed/inference/engine.py#L611
Not sure about FT, not sure when will their model.forward will be exposed. cc @rohithkrn  

In these cases, we might not need LMBlock itself. We could get the model and kwargs directly in SeqBatchScheduler constructor and call the model.forward directly for now.  (just thinking aloud)

And also do we need PtBlock and OrtBlock for Python engine?

Comment (sindhuvahinis) [reply engines/python/setup/djl_python/scheduler/lm_block.py:66]: I had a discussion with @zachgk. One more thing to consider is that, this LMBlock tries to standardize the forward function but it also loads the model. In future if we introduce more models, we could use AutoModel to load, but for Peft models, the model loading by loading the base model first and then their lora updated parameters like below. 
```
peft_model_id = "mrm8488/Alpacoom"
config = PeftConfig.from_pretrained(peft_model_id)
model = AutoModelForCausalLM.from_pretrained(config.base_model_name_or_path, return_dict=True, load_in_8bit=True, device_map={"":0})
model = PeftModel.from_pretrained(model, peft_model_id)
```
Since model loading can be arbitrary as well, I am not sure whether we should have the model loading tied to the LMBlock. I don't have any hard decision comment on this, just laying out the different sides of facts.

File: engines/python/setup/setup.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ def run(self):
 
     requirements = ['psutil', 'packaging', 'wheel']
 
-    test_requirements = ['numpy', 'requests', 'Pillow']
+    test_requirements = ['numpy', 'requests', 'Pillow', 'transformers', 'torch']
```
Comment (frankfliu) [inline engines/python/setup/setup.py:47]: I don't think Python engine should depends on torch or transformers. DJL Python engine should be generic.

One option could be create another pip package and install only in DLC.

File: engines/python/setup/djl_python/scheduler/lm_block.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
```
Comment (frankfliu) [inline engines/python/setup/djl_python/scheduler/lm_block.py:66]: Typo: HuggingfaceGPT2Block

File: engines/python/setup/djl_python/scheduler/lm_block.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    def __init__(self):
+        self.num_attention_heads = 12
+        self.num_layers = 12
+        self.hidden_state_dim = 768
+        self.logits_dim = 50257
+        self.kv_dim = 64
+
+
+class LMBlock(ABC):
```
Comment (frankfliu) [inline engines/python/setup/djl_python/scheduler/lm_block.py:31]: Are we going to have different wrapper for each model architecture?

File: engines/python/setup/djl_python/scheduler/lm_block.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    @abstractmethod
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def forward(self, input, past_key_values):
+        pass
+
+
+class PtLMBlock(LMBlock):
```
Comment (frankfliu) [inline engines/python/setup/djl_python/scheduler/lm_block.py:42]: I assume we only support Huggingface now. Let's remove Pt and Ort.

File: engines/python/setup/setup.py (hunk 3 above)
Comment (sindhuvahinis) [reply engines/python/setup/setup.py:47]: But these are for test requirements, for running unit test cases.
